core/src/main/java/jenkins/model/PeepholePermalink.java

```diff
@@ -94,115 +90,216 @@ protected File getPermalinkFile(Job<?, ?> job) {
-        } else {
-            b = null;
         }
 
-        // the cache is stale. start the search
-        if (b == null) {
-            b = job.getNearestOldBuild(n);
+        Unknown UNKNOWN = new Unknown();
+
+        /**
+         * The cache entry for this target is present.
+         */
+        sealed interface Known extends PermalinkTarget {}
+
+        /** There is known to be no matching build. */
+        record None() implements Known {
+            @Override
+            public Run<?, ?> resolve(PeepholePermalink pp, Job<?, ?> job, String id) {
+                return null;
+            }
         }
 
-        if (b == null) {
-            // no cache
-            b = job.getLastBuild();
+        /** Singleton of {@link None}. */
+        None NONE = new None();
+
+        /** A matching build, indicated by {@link Run#getNumber}. */
+        record Some(int number) implements Known {
+            @Override
+            public Run<?, ?> resolve(PeepholePermalink pp, Job<?, ?> job, String id) {
+                Run<?, ?> b = job.getBuildByNumber(number);
+                if (b != null && pp.apply(b)) {
+                    return b; // found it (in the most efficient way possible)
+                }
+                // the cache is stale. start the search
+                if (b == null) {
+                    b = job.getNearestOldBuild(number);
+                }
+                return search(pp, job, id, b);
+            }
         }
 
-        // start from the build 'b' and locate the build that matches the criteria going back in time
-        b = find(b);
+        /**
+         * Looks for any existing cache hit.
+         * @param id {@link #getId}
+         * @return {@link Some} or {@link #NONE} or {@link #UNKNOWN}
+         */
+        @NonNull PermalinkTarget get(@NonNull Job<?, ?> job, @NonNull String id);
 
-        updateCache(job, b);
-        return b;
+        /**
+         * Updates the cache.
+         * Note that this may be called not just when a build completes or is deleted
+         * (meaning that the logical value of the cache has changed),
+         * but also when {@link #resolve} has failed to find a cached value
+         * (or determined that a previously cached value is in fact invalid).
+         * @param id {@link #getId}
+         * @param target {@link Some} or {@link #NONE}
+         */
+        void put(@NonNull Job<?, ?> job, @NonNull String id, @NonNull Known target);
     }
 
     /**
-     * Start from the build 'b' and locate the build that matches the criteria going back in time
+     * Default cache based on a {@code permalinks} file in the build directory.
+     * There is one line per cached permalink, in the format {@code lastStableBuild 123}
+     * or (for a negative cache) {@code lastFailedBuild -1}.
      */
-    private Run<?, ?> find(Run<?, ?> b) {
-        //noinspection StatementWithEmptyBody
-        for ( ; b != null && !apply(b); b = b.getPreviousBuild())
-            ;
```

Comment:
The Git diff unfortunately fails to associate this hunk with the original method call. I just converted the body into a `while` loop for clarity.